### PR TITLE
Introduce signable, signature, pubkey, verifier, signer, and certific…

### PR DESCRIPTION
**Pull request by CoderZhi**
Also add uint test for these interfaces

### CODE REVIEW

File: crypto/certificate.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	if signature := c.Signature(pkey); signature != nil {
+		return verifier.Verify(signature)
+	}
+
+	return false
+}
+
+// VerifiedByAll checks whether the certificate contains the signatures of all these verifiers
+// and only these verifiers
+func (c *Certificate) VerifiedByAll(verifiers []Verifier) bool {
```
Comment (yutongp) [inline crypto/certificate.go]: can we combine `VerifiedBy` and `VerifiedByAll` to `VerifiedBy(verifiers ...Verifier)`?

File: crypto/certificate.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+// Signature defines an interface of
+type Signature interface {
+	Bytes() []byte
+}
+
+// Verifier defines an interface which could verify a signature
+type Verifier interface {
+	Verify(signature Signature) bool
```
Comment (zjshen14) [inline crypto/certificate.go]: Why not Verify(sig, pk)?

File: crypto/certificate.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	PublicKey() PublicKey
+}
+
+// Signer defines an interface which could sign a signable
+type Signer interface {
+	Verifier
+	Sign(Signable) (Signature, error)
```
Comment (zjshen14) [inline crypto/certificate.go]: Sign(signable, sk)

File: crypto/certificate.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+// Verifier defines an interface which could verify a signature
+type Verifier interface {
+	Verify(signature Signature) bool
+	PublicKey() PublicKey
+}
+
+// Signer defines an interface which could sign a signable
+type Signer interface {
+	Verifier
```
Comment (zjshen14) [inline crypto/certificate.go]: Decouple?

File: crypto/certificate.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	Sign(Signable) (Signature, error)
+}
+
+// Certificate represents a certificate for a signable
+type Certificate struct {
+	signable   Signable
```
Comment (zjshen14) [inline crypto/certificate.go]: Instead of signable, how about hash Hash256/160

File: crypto/certificate.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	return nil
+}
+
+// Endorsers returns the list of endorsers of this certificate
+func (c *Certificate) Endorsers() []PublicKey {
+	return c.endorsers
+}
+
+// Signature returns the signature of the verifier saved in this certificate
+func (c *Certificate) Signature(pk PublicKey) Signature {
```
Comment (yutongp) [inline crypto/certificate.go]: maybe `func (c *Certificate) Signature(pk PublicKey) (Signature, bool)`?

File: crypto/certificate.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+// Signer defines an interface which could sign a signable
+type Signer interface {
+	Verifier
+	Sign(Signable) (Signature, error)
+}
+
+// Certificate represents a certificate for a signable
+type Certificate struct {
```
Comment (zjshen14) [inline crypto/certificate.go]: Is it better one sig/pk per Certificate?

File: crypto/certificate.go (hunk 2 above)
Comment (CoderZhi) [reply crypto/certificate.go]: I assumed that the interface defined here contains the public key. What you suggested above is more like the interface of crypto algorithm

File: crypto/certificate.go (hunk 3 above)
Comment (CoderZhi) [reply crypto/certificate.go]: same here

File: crypto/certificate.go (hunk 4 above)
Comment (CoderZhi) [reply crypto/certificate.go]: will do

File: crypto/certificate.go (hunk 7 above)
Comment (CoderZhi) [reply crypto/certificate.go]: In most cases, we will store a list of signatures for a given signable, thus the signable will be redundant data if we store one sig/pk per certificate.

**Pull request closed without merging**